src/main/java/teammates/common/util/Url.java

```diff
@@ -17,84 +33,26 @@ protected String getAppUrl() {
-            startIndicationLocation = urlString.indexOf(startIndicator);
-        }
-
-        if (startIndicationLocation < 0) {
-            return null;
-        }
-
-        int startIndex = startIndicationLocation + parameterName.length() + 2;
-        String prefixStripped = urlString.substring(startIndex);
-        int endIndex = prefixStripped.indexOf('&');
-        if (endIndex > 0) {
-            return prefixStripped.substring(0, endIndex);
-        } else {
-            return prefixStripped;
-        }
-    }
-
-    @SuppressWarnings("unchecked")
-    public <T extends Url> T withUserId(String userId) {
-        this.urlString = addParamToUrl(this.urlString, Const.ParamsNames.USER_ID, userId);
-        return (T) this;
-    }
-
-    @SuppressWarnings("unchecked")
-    public <T extends Url> T withRegistrationKey(String key) {
-        this.urlString = addParamToUrl(this.urlString, Const.ParamsNames.REGKEY, key);
-        return (T) this;
-    }
-
-    @SuppressWarnings("unchecked")
-    public <T extends Url> T withInstructorInstitution(String institute) {
-        this.urlString = addParamToUrl(this.urlString, Const.ParamsNames.INSTRUCTOR_INSTITUTION, institute);
-        return (T) this;
-    }
-
-    @SuppressWarnings("unchecked")
-    public <T extends Url> T withCourseId(String courseId) {
-        this.urlString = addParamToUrl(this.urlString, Const.ParamsNames.COURSE_ID, courseId);
-        return (T) this;
-    }
-
-    @SuppressWarnings("unchecked")
-    public <T extends Url> T withSessionName(String feedbackSessionName) {
-        this.urlString = addParamToUrl(this.urlString, Const.ParamsNames.FEEDBACK_SESSION_NAME, feedbackSessionName);
-        return (T) this;
-    }
-
-    @SuppressWarnings("unchecked")
-    public <T extends Url> T withStudentEmail(String email) {
-        this.urlString = addParamToUrl(this.urlString, Const.ParamsNames.STUDENT_EMAIL, email);
-        return (T) this;
-    }
-
-    @SuppressWarnings("unchecked")
-    public <T extends Url> T withInstructorId(String instructorId) {
-        this.urlString = addParamToUrl(this.urlString, Const.ParamsNames.INSTRUCTOR_ID, instructorId);
-        return (T) this;
-    }
-
-    @SuppressWarnings("unchecked")
-    public <T extends Url> T withCourseName(String courseName) {
-        this.urlString = addParamToUrl(this.urlString, Const.ParamsNames.COURSE_NAME, courseName);
-        return (T) this;
+        /*
+         * Regex meaning: from the start of the string, try to find either:
+         * 1. "?" followed by "{parameterName}="
+         * 2. Any amount of any character followed by "&{parameterName}="
+         * followed by as many characters as possible until the first & is found.
+         * Returns the first occurrence if found, null otherwise.
+         */
+        String keyValuePairRegex = "^(\\?|.*&)" + parameterName + "=([^&]*).*";
+        return query.matches(keyValuePairRegex) ? query.replaceFirst(keyValuePairRegex, "$2") : null;
```

Comment:
Java's URL object doesn't have such capability. It can tell which part of the URL is query but not anything more specific. This is already the parent class.